### DISCUSSION TITLE
Question about zstd stateless decoding

### DISCUSSION THREAD

**Discussion by luxflow** [klauspost/compress#995] [Q&A] [answered]
I have read through the documentation but still do not fully understand the differences between stream and stateless decoding. Could you provide a reference document or an explanation?

As I understand that when decoding in a stream mode, only one stream can be decoded at a time. 
When I request a zstd-encoded HTTP page multiple times, Is there a possibility of bottlenecks occurring due to decoding compared to stateless mode?

When requesting a zstd-encoded HTTP page, what are the differences between decoding it using stream mode and receiving the stream completely and decoding buffer in stateless mode?
Is there a performance or compression ratio difference between the two?
I plan to use the stateless mode for ~10K requests within an intranet, with esponse sizes < 1MB, and response time < 1s
In this scenario, would it be appropriate to use the stateless method?

thank you in advance

**Comment by klauspost** [comment, marked as answer]
The main difference is the memory requirements, but also some speed considerations.

The []byte -> []byte requires you to have both the input and the output in memory at the same time. If your payload is a 1GB JSON file, this is probably not a great idea.  

Secondly []byte -> []byte means that every step must complete before the next can start. This means your flow will be Request -> io.ReadAll -> DecodeAll -> Process serial flow. With a stream, as soon as the first block is returned by the remote and has been compressed it is ready for your processor.

However streams require more setup, so typically for smaller payloads it is not worth setting it up, for effectively a single block (up to 128KB) of data will provide none of these benefits.

**Reply by luxflow** [reply]
I'm sorry for posting question to issues
Your answer helps my understanding thank you!
